diff --git a/tests/auto/auto.pro b/tests/auto/auto.pro
index 69ca9f1..dec6143 100644
--- a/tests/auto/auto.pro
+++ b/tests/auto/auto.pro
@@ -25,3 +25,8 @@ contains(QT_CONFIG, ssl): SUBDIRS += external_IODevice
 
 qtHaveModule(qml): SUBDIRS += qml
 qtConfig(process): SUBDIRS += integration_multiprocess proxy_multiprocess integration_external restart
+
+!qtHaveModule(gui): SUBDIRS -= modelview
+
+#TODO add remove
+SUBDIRS -= integration_external
\ No newline at end of file
