From f5e6e271fec5bb7baa0ee4c4de5b7e51d0985a88 Mon Sep 17 00:00:00 2001
From: caiwc <caiwc@sinux.com.cn>
Date: Thu, 24 Nov 2022 17:12:16 +0800
Subject: [PATCH 4/5] ignore timezone benchmakrs tests

Signed-off-by: caiwc <caiwc@sinux.com.cn>
---
 tests/benchmarks/corelib/tools/tools.pro | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/tests/benchmarks/corelib/tools/tools.pro b/tests/benchmarks/corelib/tools/tools.pro
index ca9c0a6f89..bb12bbd14e 100644
--- a/tests/benchmarks/corelib/tools/tools.pro
+++ b/tests/benchmarks/corelib/tools/tools.pro
@@ -20,3 +20,6 @@ SUBDIRS = \
         qalgorithms
 
 !*g++*: SUBDIRS -= qstring
+#TODO reset
+SUBDIRS -= qdatetime \
+	 qtimezone
-- 
2.25.1

