diff --git a/.gitignore b/.gitignore
index 9b8015da..c8f4c823 100644
--- a/.gitignore
+++ b/.gitignore
@@ -2,3 +2,17 @@ Makefile
 .qmake.cache
 .qmake.super
 *.pro.user*
+.vscode
+.qmake.stash
+config.cache
+config.log
+config.opt
+config.status.bat
+config.summary
+*/config.tests/*
+*.o
+*.obj
+*.moc
+moc_*.cpp
+*-debug/
+*-release/
\ No newline at end of file
diff --git a/.gitlab-ci.yml b/.gitlab-ci.yml
new file mode 100644
index 00000000..805db031
--- /dev/null
+++ b/.gitlab-ci.yml
@@ -0,0 +1,182 @@
+stages:
+  - HarmonyOS-17
+  - HarmonyOS-16
+  - HarmonyOS-15
+  - HarmonyOS-14
+  - HarmonyOS-13
+  - OpenHarmony-15
+
+
+variables:
+  GIT_SUBMODULE_STRATEGY: recursive
+
+###################### Qt5.15.12 ######################
+
+Qt5_15_12-OHOS17 for Windows:
+  stage: HarmonyOS-17
+  variables:
+    BUILD_TYPE: profile
+    OHOS_ARCH: arm64-v8a
+    SELECT_OHOS_API: 17
+  script:
+    - .\issbuild\harmonyos\windows_build.bat
+  tags:
+    - Windows_x86_64_Shared_Runner
+  only:
+    - 5.15.12-ohos-next
+  
+Qt5_15_12-OHOS16 for Windows:
+  stage: HarmonyOS-16
+  variables:
+    BUILD_TYPE: profile
+    OHOS_ARCH: arm64-v8a
+    SELECT_OHOS_API: 16
+  script:
+    - .\issbuild\harmonyos\windows_build.bat
+  tags:
+    - Windows_x86_64_Shared_Runner
+  only:
+    - 5.15.12-ohos-next
+
+Qt5_15_12-OHOS15 for Windows:
+  stage: HarmonyOS-15
+  variables:
+    BUILD_TYPE: profile
+    OHOS_ARCH: arm64-v8a
+    SELECT_OHOS_API: 15
+  script:
+    - .\issbuild\harmonyos\windows_build.bat
+  tags:
+    - Windows_x86_64_Shared_Runner
+  only:
+    - 5.15.12-ohos-next
+
+.Qt5_15_12-OHOS14 for Windows:
+  stage: HarmonyOS-14
+  variables:
+    BUILD_TYPE: profile
+    OHOS_ARCH: arm64-v8a
+    SELECT_OHOS_API: 14
+  script:
+    - .\issbuild\harmonyos\windows_build.bat
+  tags:
+    - Windows_x86_64_Shared_Runner
+  only:
+    - 5.15.12-ohos-next
+
+Qt5_15_12-OHOS17 for Linux:
+  stage: HarmonyOS-17
+  variables:
+    BUILD_TYPE: profile
+    OHOS_ARCH: arm64-v8a
+  script:
+    - chmod +x ./issbuild/harmonyos/*
+    - ./issbuild/harmonyos/linux_build.sh
+  tags:
+    - Linux_x86_64_Shared_Runner
+  image:
+    name: liyaosong/ohos-sdk:17
+  only:
+    - 5.15.12-ohos-next
+
+Qt5_15_12-OHOS16 for Linux:
+  stage: HarmonyOS-16
+  variables:
+    BUILD_TYPE: profile
+    OHOS_ARCH: arm64-v8a
+  script:
+    - chmod +x ./issbuild/harmonyos/*
+    - ./issbuild/harmonyos/linux_build.sh
+  tags:
+    - Linux_x86_64_Shared_Runner
+  image:
+    name: liyaosong/ohos-sdk:16
+  only:
+    - 5.15.12-ohos-next
+
+Qt5_15_12-OHOS15 for Linux:
+  stage: HarmonyOS-15
+  variables:
+    BUILD_TYPE: profile
+    OHOS_ARCH: arm64-v8a
+  script:
+    - chmod +x ./issbuild/harmonyos/*
+    - ./issbuild/harmonyos/linux_build.sh
+  tags:
+    - Linux_x86_64_Shared_Runner
+  image:
+    name: liyaosong/ohos-sdk:15
+  only:
+    - 5.15.12-ohos-next
+
+.Qt5_15_12-OHOS14 for Linux:
+  stage: HarmonyOS-14
+  variables:
+    BUILD_TYPE: profile
+    OHOS_ARCH: arm64-v8a
+  script:
+    - chmod +x ./issbuild/harmonyos/*
+    - ./issbuild/harmonyos/linux_build.sh
+  tags:
+    - Linux_x86_64_Shared_Runner
+  image:
+    name: liyaosong/ohos-sdk:14
+  only:
+    - 5.15.12-ohos-next
+  
+######################## OpenHarmony-15 ######################
+Qt5_15_12-OpenHarmony15 for Windows:
+  stage: OpenHarmony-15
+  variables:
+    BUILD_TYPE: profile
+    OHOS_ARCH: arm64-v8a
+    SELECT_OHOS_API: 15
+  script:
+    - .\issbuild\openharmony\windows_build.bat
+  tags:
+    - Windows_x86_64_Shared_Runner
+  only:
+    - 5.15.12-openharmony
+
+Qt5_15_12-OpenHarmony15-arm32 for Windows:
+  stage: OpenHarmony-15
+  variables:
+    BUILD_TYPE: profile
+    OHOS_ARCH: armeabi-v7a
+    SELECT_OHOS_API: 15
+  script:
+    - .\issbuild\openharmony\windows_build.bat
+  tags:
+    - Windows_x86_64_Shared_Runner
+  only:
+    - 5.15.12-openharmony
+
+.Qt5_15_12-OpenHarmony15 for Linux:
+  stage: OpenHarmony-15
+  variables:
+    BUILD_TYPE: profile
+    OHOS_ARCH: arm64-v8a
+  script:
+    - chmod +x ./issbuild/openharmony/*
+    - ./issbuild/openharmony/linux_build.sh
+  tags:
+    - Linux_x86_64_Shared_Runner
+  image:
+    name: liyaosong/ohos-sdk:15
+  only:
+    - 5.15.12-openharmony
+
+.Qt5_15_12-OpenHarmony15-arm32 for Linux:
+  stage: OpenHarmony-15
+  variables:
+    BUILD_TYPE: profile
+    OHOS_ARCH: armeabi-v7a
+  script:
+    - chmod +x ./issbuild/openharmony/*
+    - ./issbuild/openharmony/linux_build.sh
+  tags:
+    - Linux_x86_64_Shared_Runner
+  image:
+    name: liyaosong/ohos-sdk:15
+  only:
+    - 5.15.12-openharmony
\ No newline at end of file
diff --git a/.gitmodules b/.gitmodules
index 45223365..43cb70c1 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -1,7 +1,7 @@
 [submodule "qtbase"]
 	path = qtbase
 	url = ../qtbase.git
-	branch = 5.15
+	branch = 5.15.17-openharmony
 	status = essential
 [submodule "qtsvg"]
 	depends = qtbase
@@ -14,7 +14,7 @@
 	recommends = qtsvg
 	path = qtdeclarative
 	url = ../qtdeclarative.git
-	branch = 5.15
+	branch = 5.15.17-openharmony
 	status = essential
 [submodule "qtactiveqt"]
 	depends = qtbase
@@ -34,7 +34,7 @@
 	recommends = qtdeclarative
 	path = qtmultimedia
 	url = ../qtmultimedia.git
-	branch = 5.15
+	branch = 5.15.17-openharmony
 	status = essential
 [submodule "qttools"]
 	depends = qtbase
@@ -90,7 +90,7 @@
 	recommends = qtdeclarative
 	path = qtsensors
 	url = ../qtsensors.git
-	branch = 5.15
+	branch = 5.15.17-openharmony
 	status = addon
 [submodule "qtsystems"]
 	depends = qtbase
@@ -123,7 +123,7 @@
 	recommends = qtdeclarative qtandroidextras
 	path = qtconnectivity
 	url = ../qtconnectivity.git
-	branch = 5.15
+	branch = 5.15.17-openharmony
 	status = addon
 [submodule "qtwayland"]
 	depends = qtbase
@@ -156,7 +156,7 @@
 	recommends = qtgraphicaleffects
 	path = qtquickcontrols
 	url = ../qtquickcontrols.git
-	branch = 5.15
+	branch = 5.15.17-openharmony
 	status = addon
 [submodule "qtserialbus"]
 	depends = qtbase
@@ -315,7 +315,7 @@
 [submodule "qtquicktimeline"]
 	depends = qtbase qtdeclarative
 	path = qtquicktimeline
-	url = ../qtquicktimeline
+	url = ../qtquicktimeline.git
 	branch = 5.15
 	status = addon
 [submodule "qtquick3d"]
@@ -324,3 +324,19 @@
 	url = ../qtquick3d.git
 	branch = 5.15
 	status = addon
+[submodule "issbuild"]
+	path = issbuild
+	url = ../issbuild.git
+	status = addon
+	branch = main
+[submodule "qtohextras"]
+	depends = qtbase
+	path = qtohextras
+	url = ../qtohextras.git
+	status = addon
+	branch = 5.15.17-openharmony
+[submodule "qtpdf"]
+	depends = qtbase
+	path = qtpdf
+	url = ../qtpdf.git
+	branch = 5.15.17-openharmony
diff --git a/issbuild b/issbuild
new file mode 160000
index 00000000..042513ba
--- /dev/null
+++ b/issbuild
@@ -0,0 +1 @@
+Subproject commit 042513baa292cb38b5692a5d202d58ae3f12a4cf
diff --git a/qtbase b/qtbase
index 8e79bee4..46269e4c 160000
--- a/qtbase
+++ b/qtbase
@@ -1 +1 @@
-Subproject commit 8e79bee4afa2a1466f360f44fb07d24e432a82a6
+Subproject commit 46269e4c111849993e787c16849dd8a2d1c86922
diff --git a/qtconnectivity b/qtconnectivity
index e8de0149..43304642 160000
--- a/qtconnectivity
+++ b/qtconnectivity
@@ -1 +1 @@
-Subproject commit e8de014972fc254c1125a4da24597009f2f6cc68
+Subproject commit 43304642b1c715178d1df936f44dee56ec967b88
diff --git a/qtdeclarative b/qtdeclarative
index 8fb72927..10e040b3 160000
--- a/qtdeclarative
+++ b/qtdeclarative
@@ -1 +1 @@
-Subproject commit 8fb7292711044cc256b8aaf02913ce0cd79d0d21
+Subproject commit 10e040b3780582f97ce199e399c2746a2ccb4e58
diff --git a/qtmultimedia b/qtmultimedia
index 09dfcd32..a0a05bfb 160000
--- a/qtmultimedia
+++ b/qtmultimedia
@@ -1 +1 @@
-Subproject commit 09dfcd32f0255d6844a6bc16a1ddb9fd5bacdb7b
+Subproject commit a0a05bfb6e6bf0bef9164ec74e17fe70b1c10327
diff --git a/qtohextras b/qtohextras
new file mode 160000
index 00000000..7a57d49e
--- /dev/null
+++ b/qtohextras
@@ -0,0 +1 @@
+Subproject commit 7a57d49e35325eb25681a8b832a72c9a6a5319dc
diff --git a/qtpdf b/qtpdf
new file mode 160000
index 00000000..c79ef4b9
--- /dev/null
+++ b/qtpdf
@@ -0,0 +1 @@
+Subproject commit c79ef4b92f706b94a75fb4d8a15caf2d42de51e4
diff --git a/qtquickcontrols b/qtquickcontrols
index 37ee5e07..d48805ef 160000
--- a/qtquickcontrols
+++ b/qtquickcontrols
@@ -1 +1 @@
-Subproject commit 37ee5e07e82296e24cc02c8bf6f31cb6bd8846b1
+Subproject commit d48805efedb12980781c007b8d0b635697245430
diff --git a/qtsensors b/qtsensors
index 79ed5db9..5ede4973 160000
--- a/qtsensors
+++ b/qtsensors
@@ -1 +1 @@
-Subproject commit 79ed5db98f36b8ca70b077edef46fc793debb3ab
+Subproject commit 5ede49735211fad86b67dda411f46377e2c3fedd
